Adds procedure to create private cluster with private image registry (#69087)
openshift/openshift-docs: modules/installing-private-image-registry-private-azure.adoc

      - Microsoft.Storage/storageAccounts/blobServices/containers/write
      - Microsoft.Storage/storageAccounts/blobServices/generateUserDelegationKey/action
      - Microsoft.Storage/storageAccounts/read
      - Microsoft.Storage/storageAccounts/write
      - Microsoft.Storage/storageAccounts/delete
      - Microsoft.Storage/storageAccounts/listKeys/action
      - Microsoft.Resources/tags/write
      - Microsoft.Network/privateEndpoints/write
      - Microsoft.Network/privateEndpoints/read
      - Microsoft.Network/privateEndpoints/privateDnsZoneGroups/write
      - Microsoft.Network/privateEndpoints/privateDnsZoneGroups/read
      - Microsoft.Network/privateDnsZones/read
      - Microsoft.Network/privateDnsZones/write
      - Microsoft.Network/privateDnsZones/join/action
      - Microsoft.Network/privateDnsZones/A/write
      - Microsoft.Network/privateDnsZones/virtualNetworkLinks/write
      - Microsoft.Network/privateDnsZones/virtualNetworkLinks/read
      - Microsoft.Network/virtualNetworks/subnets/join/action
      - Microsoft.Network/virtualNetworks/join/action
      - Microsoft.Network/networkInterfaces/read

>       - Microsoft.Storage/storageAccounts/PrivateEndpointConnectionsApproval/action

Reviewer: I think those permissions should not be listed in the doc, but only adding into registry credential request manifest, which will be consumed when creating separated credentials if cluster is configured in credentialsMode: Manual mode.

Author: @jinyunma To be sure, you're requesting that I remove the permissions from this procedure? I have updated the PR to remove this step. 

If an alternative step is required, would you mind providing it?

Reviewer: I think no alternative step is required, cc @flavianmissi and @wewang58 to check or any other concern?

Reviewer 2: Yes no alternative step is required

+
[source,terminal]
----
$ touch imageregistry-config.yaml
----
+
[source,yaml]
----
apiVersion: imageregistry.operator.openshift.io/v1
kind: Config
metadata:
  name: cluster
spec:
  managementState: "Managed"
  replicas: 2
  rolloutStrategy: RollingUpdate
  storage:
    azure:
      networkAccess:
        internal:
          networkResourceGroupName: <vnet_resource_group> <1>
          subnetName: <subnet_name> <2>
          vnetName: <vnet_name> <3>
        type: Internal 
----
<1> Replace `<vnet_resource_group>` with the resource group name that contains the existing virtual network (VNet).
<2> Replace `<subnet_name>` with the name of the existing subnet within the specified resource group.
<3> Replace `<vnet_name>` with the name of the existing virtual network (VNet) in the specified resource group.

. Move the `imageregistry-config.yaml` file to the `<installation_directory/manifests>` folder by running the following command:
+
[source,terminal]
----
$ mv imageregistry-config.yaml <installation_directory/manifests/>
----

.Next steps

> * After you have moved the `imageregistry-config.yaml` file to the `<installation_directory/manifests>` folder and set the required permissions, proceed to "Deploying the cluster". You can also ensure that a private endpoint resource has been generated by following the _Verification_ steps found in the same section.

Reviewer: 1. I think user can still continue the following procedures ( providing short-term or long-term credentials when setting credentialsMode: Manual) instead of proceeding to "Deploying the cluster" directly, right?
2.  what does the _Verification_ steps here refer to?

Reviewer: I found _Verificaiton_ step, but seems that it is present in all azure installation doc, this step should only be in private cluster, right?

Author: Yes. Fixed. Thanks @jinyunma

// Module included in the following assemblies:
//
//* registry/configuring_registry_storage-azure.adoc

:_mod-docs-content-type: PROCEDURE
[id="installing-private-image-registry-private-azure"]
= Optional: Preparing a private Microsoft Azure cluster for a private image registry

By installing a private image registry on a private Microsoft Azure cluster, you can create private endpoints. Private endpoints disable public facing endpoints to the internal registry, adding an extra layer of security to your {product-title} deployment.

Use the following guide to prepare your private Microsoft Azure cluster for installation with a private image registry.

.Prerequisites

* You have access to an {product-title} account with cluster administrator access.

* You have installed the OpenShift CLI (oc).

* You have prepared an `install-config.yaml` that includes the following information:

> ** A file that includes your VPC configuration

Reviewer 3: what file is this? I thought networking (vpc) configuration was part of the install config? 🤔

Author: I really don't remember where this came from. I've removed the following line:

* A file that includes your VPC configuration

If the line is required I will re-add it.

----
$ ./openshift-install create manifests --dir <installation_directory>
----
+
This command displays the following messages:
+
.Example output
[source,terminal]
----
INFO Consuming Install Config from target directory
INFO Manifests created in: <installation_directory>/manifests and <installation_directory>/openshift
----

. Create an image registry configuration object and pass in the `networkResourceGroupName`, `subnetName`, and `vnetName` provided by Microsoft Azure. For example:
+
[source,terminal]
----
$ touch imageregistry-config.yaml
----
+
[source,yaml]
----
apiVersion: imageregistry.operator.openshift.io/v1
kind: Config
metadata:
  name: cluster
spec:
  managementState: "Managed"
  replicas: 2
  rolloutStrategy: RollingUpdate
  storage:
    azure:
      networkAccess:
        internal:
          networkResourceGroupName: <vnet_resource_group> <1>
          subnetName: <subnet_name> <2>
          vnetName: <vnet_name> <3>
        type: Internal 
----
<1> Replace `<vnet_resource_group>` with the resource group name that contains the existing virtual network (VNet).
<2> Replace `<subnet_name>` with the name of the existing subnet within the specified resource group.

> <3> Replace `<vnet_name>` with the name of the existing virtual network (VNet) in the specified resource group.

Reviewer 3: is it worth mentioning that 1, 2 and 3 are optional if the user doesn't have an existing vnet/subnet setup?

Author: Fixed.

> . Move the `imageregistry-config.yaml` file to the `<installation_directory/manifests>` folder by running the following command:

Reviewer 3: don't know if it matters here but this file is is consumed after installation.

Author: Add a note:

[NOTE]
====
The `imageregistry-config.yaml` file is consumed during the installation process. If desired, you must back it up before installation.
====

// Module included in the following assemblies:
//
//* registry/configuring_registry_storage-azure.adoc

:_mod-docs-content-type: PROCEDURE
[id="installing-private-image-registry-private-azure"]
= Optional: Preparing a private Microsoft Azure cluster for a private image registry

By installing a private image registry on a private Microsoft Azure cluster, you can create private storage endpoints. Private storage endpoints disable public facing endpoints to the registry's storage account, adding an extra layer of security to your {product-title} deployment.
Use the following guide to prepare your private Microsoft Azure cluster for installation with a private image registry.

.Prerequisites

> * You have access to an {product-title} account with cluster administrator access.

Reviewer 4: I think that you can remove the "You have...". It's implied.

// Module included in the following assemblies:
//
//* registry/configuring_registry_storage-azure.adoc

:_mod-docs-content-type: PROCEDURE
[id="installing-private-image-registry-private-azure"]
= Optional: Preparing a private Microsoft Azure cluster for a private image registry

> By installing a private image registry on a private Microsoft Azure cluster, you can create private endpoints. Private endpoints disable public facing endpoints to the internal registry, adding an extra layer of security to your {product-title} deployment.

Reviewer 3: Let's also use `private storage endpoints` everywhere on this PR instead of just `private endpoints` (so there's no chance of confusing them with route endpoints)
```suggestion
By installing a private image registry on a private Microsoft Azure cluster, you can create private storage endpoints. Private storage endpoints disable public facing endpoints to the registry's storage account, adding an extra layer of security to your {product-title} deployment.
```

Author: Fixed!

Use the following guide to prepare your private Microsoft Azure cluster for installation with a private image registry.

.Prerequisites

* You have access to an {product-title} account with cluster administrator access.

* You have installed the OpenShift CLI (oc).

* You have prepared an `install-config.yaml` that includes the following information:
** The `publish` field is set to `Internal`

> * You have created a global Microsoft Azure account for the `ccoctl` utility.

Reviewer: Here might be updated to indicate Azure account which have permission including https://github.com/openshift/openshift-docs/pull/69087/files#diff-c82de66f7c129811e1de419d2c8b04c6cf8429d036db2acffc36f43e606fdcbaR149? but not for "ccoctl" utility.

> * You have set the permissions for creating an image registry storage account private endpoint. For more information, see "Azure permissions for installer-provisioned infrastructure".

Reviewer 3: `image registry storage account private endpoint` is a mouth full. I think we can use just `private storage endpoint` or something in these lines. WDYT?
```suggestion
* You have set the permissions for creating a private storage endpoint. For more information, see "Azure permissions for installer-provisioned infrastructure".
```

Author: Fixed!

// Module included in the following assemblies:
//
//* registry/configuring_registry_storage-azure.adoc

:_mod-docs-content-type: PROCEDURE
[id="installing-private-image-registry-private-azure"]
= Optional: Preparing a private Microsoft Azure cluster for a private image registry

By installing a private image registry on a private Microsoft Azure cluster, you can create private endpoints. Private endpoints disable public facing endpoints to the internal registry, adding an extra layer of security to your {product-title} deployment.

Use the following guide to prepare your private Microsoft Azure cluster for installation with a private image registry.

.Prerequisites

* You have access to an {product-title} account with cluster administrator access.

* You have installed the OpenShift CLI (oc).

* You have prepared an `install-config.yaml` that includes the following information:
** The `publish` field is set to `Internal`
** The `credentialsMode` field is set to `Manual`

> * You have created a global Microsoft Azure account for the `ccoctl` utility.

Reviewer: crendentialMode setting seems not prerequisites for private image registry, it just only requires private cluster (publish:Internal).

Reviewer 2: +1

Author: removed ** The `credentialsMode` field is set to `Manual`

+
[source,terminal]
----
$ ./openshift-install create manifests --dir <installation_directory>
----
+
This command displays the following messages:
+
.Example output
[source,terminal]
----
INFO Consuming Install Config from target directory
INFO Manifests created in: <installation_directory>/manifests and <installation_directory>/openshift
----

. Create an image registry configuration object and pass in the `networkResourceGroupName`, `subnetName`, and `vnetName` provided by Microsoft Azure. For example:
+
[source,terminal]
----
$ touch imageregistry-config.yaml
----
+
[source,yaml]
----
apiVersion: imageregistry.operator.openshift.io/v1
kind: Config
metadata:
  name: cluster
spec:
  managementState: "Managed"
  replicas: 2
  rolloutStrategy: RollingUpdate
  storage:
    azure:
      networkAccess:
        internal:
          networkResourceGroupName: <vnet_resource_group> <1>
          subnetName: <subnet_name> <2>
          vnetName: <vnet_name> <3>
        type: Internal 
----
<1> Optional. If you have an existing VNet and subnet setup, replace `<vnet_resource_group>` with the resource group name that contains the existing virtual network (VNet).

> <2> Optional. If you have an existing VNet and subnet setup, replace `<subnet_name>` with the name of the existing subnet within the specified resource group.

Reviewer: Here should be compute machine's subnet, @flavianmissi correct me if I'm wrong.

Reviewer 3: good point @jinyunma, we should mention that the subnetName should be the name of the compute subnet (the alternative is control plane subnet).
i.e
```
replace `<subnet_name>` with the name of the existing compute subnet within the specified resource group
```

Author: Updated thanks @jinyunma @flavianmissi !!!!!!

// Module included in the following assemblies:
//
//* registry/configuring_registry_storage-azure.adoc

:_mod-docs-content-type: PROCEDURE
[id="installing-private-image-registry-private-azure"]
= Optional: Preparing a private Microsoft Azure cluster for a private image registry

By installing a private image registry on a private Microsoft Azure cluster, you can create private endpoints. Private endpoints disable public facing endpoints to the internal registry, adding an extra layer of security to your {product-title} deployment.

Use the following guide to prepare your private Microsoft Azure cluster for installation with a private image registry.

.Prerequisites

> * You have created a global Microsoft Azure account to use with the following permissions:

Reviewer: Here should be Azure service principal or managed identity instead of "global Microsoft Azure account".

How about updating something as below, no need to mention account, as it has already been mentioned in [Required Azure permissions for installer-provisioned infrastructure](https://69087--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_azure/installing-azure-account#minimum-required-permissions-ipi-azure_installing-azure-account).

`The following permissions are required for configuring internal image registry:`

+

> .Required permissions for creating a private image registry endpoint

Reviewer: @flavianmissi @wewang58  when using azure workload identity, those permissions should also be required on identity for image-registry, is that right?

Author: Set to "Required"

[%collapsible]
====
* `Microsoft.Network/privateEndpoints/write`
* `Microsoft.Network/privateEndpoints/read`
* `Microsoft.Network/privateEndpoints/privateDnsZoneGroups/write`
* `Microsoft.Network/privateEndpoints/privateDnsZoneGroups/read`
* `Microsoft.Network/privateDnsZones/join/action`

>  `Microsoft.Storage/storageAccounts/PrivateEndpointConnectionsApproval/action`

Reviewer: seems missing '*' in this line?
